add an API for external backends to register custom device names

bdhirsh opened the pull request:
This API adds some improvements to external backends who are building C++ backends out of tree using the `PrivateUse1` dispatch key.

The docs and linked examples go over the API in more detail, but you should be able to use it like:
```
# This should probably be in the __init__.py file of a external backend's python package
> torch.register_privateuse1_backend("foo")`
# And it will allow the user to do this:
> a = torch.ones(2, device="foo")
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86992

bdhirsh committed:
add an API for external backends to register custom device names

[ghstack-poisoned]

bdhirsh committed:
Update on "add an API for external backends to register custom device names"

This API adds some improvements to external backends who are building C++ backends out of tree using the `PrivateUse1` dispatch key.

The docs and linked examples go over the API in more detail, but you should be able to use it like:
```
# This should probably be in the __init__.py file of a external backend's python package
> torch.register_privateuse1_backend("foo")`
# And it will allow the user to do this:
> a = torch.ones(2, device="foo")
```




[ghstack-poisoned]

bdhirsh committed:
Update on "add an API for external backends to register custom device names"

This API adds some improvements to external backends who are building C++ backends out of tree using the `PrivateUse1` dispatch key.

The docs and linked examples go over the API in more detail, but you should be able to use it like:
```
# This should probably be in the __init__.py file of a external backend's python package
> torch.register_privateuse1_backend("foo")`
# And it will allow the user to do this:
> a = torch.ones(2, device="foo")
```




[ghstack-poisoned]

albanD commented on torch/_torch_docs.py:
```diff
@@ -13669,6 +13669,36 @@ def merge_dicts(*dicts):
+
+Example::
+
+    >>> torch.register_privateuse1_backend("foo_device")
```
Should be only "foo" here.

albanD commented on torch/csrc/autograd/python_torch_functions_manual.cpp:
```diff
@@ -505,6 +505,22 @@ static PyObject* THPVariable__disable_functionalization(
   END_HANDLE_TH_ERRORS
 }
 
+static PyObject* THPVariable_register_privateuse1_backend(
```
nit for future APIs like this, we can do it with pybind.
But this is good here.

bdhirsh replied:
oh good point, I mostly cargo-culted

albanD commented on torch/_torch_docs.py:
```diff
@@ -13669,6 +13669,36 @@ def merge_dicts(*dicts):
 )
 
+add_docstr(
+    torch.register_privateuse1_backend,
```
Not sure about the name here.
In particular, keeping in the back of my mind that we might need to extend this to support multiple backends.

I am wondering if we want to make an API that will support the multi backend case directly here. And thus naming it something like `torch.register_new_backend("foo")`. And then from c++ recommend they use `getDispatchKeyForDeviceType(parse_type("foo"))` (not sure what is the exact name for this?) to register their kernels.

The other option is to say they need to use PrivateUse1 manually and thus this API is just to rename it and much more limited. In such a case, I would suggest we move it to a smaller namespace like `torch.utils.rename_privateuse1_backend("foo")`.

Neither of these options require changing the way things are implemented right now btw, just the python naming, the doc and the error message when multiple backends are being used.
WDYT?

bdhirsh replied:
> And then from c++ recommend they use `getDispatchKeyForDeviceType(parse_type("foo"))`

I don't think this will easily work without some bigger API changes, since right now on the C++ side you still need to hardcode the PrivateUse1 key name directly into our registration macro: `TORCH_LIBRARY_IMPL(aten, PrivateUse1, m) {...}`.

If we wanted to fully go down the multi-backend route now (and abstract PrivateUse1 away from any public API's), one way would be with a new macro like `TORCH_BACKEND_IMPL(aten, foo, m {...})` where we set up the link between a PrivateUse key and the backend name at library load time. That would make a python API like `torch.register_new_backend()` redundant though.

LMK if you agree with that line of reasoning - but I'm leaning towards your second proposal, which is to use a smaller namespace (`torch.utils.rename_privateuse1_backend`)

albanD replied:
Sounds good!

albanD reviewed (commented):
Sounds good to me a couple questions about possible improvements.

bdhirsh committed:
Update on "add an API for external backends to register custom device names"

This API adds some improvements to external backends who are building C++ backends out of tree using the `PrivateUse1` dispatch key.

The docs and linked examples go over the API in more detail, but you should be able to use it like:
```
# This should probably be in the __init__.py file of a external backend's python package
> torch.register_privateuse1_backend("foo")`
# And it will allow the user to do this:
> a = torch.ones(2, device="foo")
```




[ghstack-poisoned]

albanD reviewed (approved):
SGTM
lint needs fixing though

bdhirsh committed:
Update on "add an API for external backends to register custom device names"

This API adds some improvements to external backends who are building C++ backends out of tree using the `PrivateUse1` dispatch key.

The docs and linked examples go over the API in more detail, but you should be able to use it like:
```
# This should probably be in the __init__.py file of a external backend's python package
> torch.register_privateuse1_backend("foo")`
# And it will allow the user to do this:
> a = torch.ones(2, device="foo")
```




[ghstack-poisoned]

bdhirsh committed:
Update on "add an API for external backends to register custom device names"

This API adds some improvements to external backends who are building C++ backends out of tree using the `PrivateUse1` dispatch key.

The docs and linked examples go over the API in more detail, but you should be able to use it like:
```
# This should probably be in the __init__.py file of a external backend's python package
> torch.register_privateuse1_backend("foo")`
# And it will allow the user to do this:
> a = torch.ones(2, device="foo")
```




[ghstack-poisoned]

bdhirsh committed:
Update on "add an API for external backends to register custom device names"

This API adds some improvements to external backends who are building C++ backends out of tree using the `PrivateUse1` dispatch key.

The docs and linked examples go over the API in more detail, but you should be able to use it like:
```
# This should probably be in the __init__.py file of a external backend's python package
> torch.register_privateuse1_backend("foo")`
# And it will allow the user to do this:
> a = torch.ones(2, device="foo")
```




[ghstack-poisoned]

bdhirsh committed:
Update on "add an API for external backends to register custom device names"

This API adds some improvements to external backends who are building C++ backends out of tree using the `PrivateUse1` dispatch key.

The docs and linked examples go over the API in more detail, but you should be able to use it like:
```
# This should probably be in the __init__.py file of a external backend's python package
> torch.register_privateuse1_backend("foo")`
# And it will allow the user to do this:
> a = torch.ones(2, device="foo")
```




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge -f "flaky failure"

jeanschmidt commented:
@pytorchbot revert -m "breaking internal builds - D40534212 - arstudio-windows-tests-landcastle-0" -c "ghfirst"

jeanschmidt commented:
Seems it is breaking windows builds:

```
stderr: xplat\caffe2\c10\core\DeviceType.cpp(139,38): error: no member named 'toupper' in namespace 'std'; did you mean simply 'toupper'?
   ->         [](unsigned char c) { return std::toupper(c); });
   ->                                      ^~~~~~~~~~~~
   ->                                      toupper
   -> C:/tools/toolchains/vs2017_15.9/WindowsSdk/Include/10.0.17763.0/ucrt\ctype.h(59,56): note: 'toupper' declared here
   -> Check_return
CRT_JIT_INTRINSIC _ACRTIMP int __cdecl toupper(_In int _C);
   ->                                                        ^
   -> xplat\caffe2\c10\core\DeviceType.cpp(157,36): error: no member named 'tolower' in namespace 'std'; did you mean simply 'tolower'?
   ->       [](unsigned char c) { return std::tolower(c); });
   ->                                    ^~~~~~~~~~~~
   ->                                    tolower
   -> C:/tools/toolchains/vs2017_15.9/WindowsSdk/Include/10.0.17763.0/ucrt\ctype.h(62,56): note: 'tolower' declared here
   -> Check_return
CRT_JIT_INTRINSIC _ACRTIMP int __cdecl tolower(_In int _C);
   ->                                                        ^
 ```

check toupper/tolower usage

bdhirsh commented:
Closing, reland at https://github.com/pytorch/pytorch/pull/87453
